### PR TITLE
Spack package for UM

### LINKED ISSUE #129: Spack package for UM
See ["Building UM executables: draft strategy"](https://forum.access-hive.org.au/t/building-um-executables-draft-strategy/2233). Create a Spack package for the UM executables `um-atmos.exe` and `um-recon.exe`, initially for UM vn13.0, as used by the Regional Nesting Suite.

### PR DESCRIPTION
**Pull request by penguian**
Closes #129
This is a minimum viable Spack package for UM to support ACCESS-RAM3.   It still uses the Subversion sources and configuration options as per `fcm_make_um` in (e.g.) [u-dg768](https://code.metoffice.gov.uk/trac/roses-u/browser/d/g/7/6/8/trunk/app/fcm_make_um). 

The intention for future development would be to:
1. Make the Spack package more configurable by introducing more variants.
2. Introduce a `model` (or similarly named) variant with an `access-ram3` option, so that the default build does not use any `diff` locations.
3. Use source code from private GitHub repositories, once they become available.
Each step would be documented in more detail via issues and pull requests in this repository.

### CODE REVIEW

File: packages/fcm/package.py [hunk 1]
```diff
@@ -8,15 +8,25 @@ class Fcm(Package):
...
+        values=("none", "nci-gadi"), multi=False)
+
     def url_for_version(self, version):
         return "https://github.com/metomi/fcm/archive/refs/tags/{0}.tar.gz".format(version)
 
     def install(self, spec, prefix):
         install_tree(".", prefix)
+        site = spec.variants["site"].value
+        if site != "none":
+            site
```
Comment (harshula) [inline packages/fcm/package.py:28]: Typo?

File: packages/um/package.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    maintainers("penguian")
+
+    variant("optim", default="safe", description="Optimization level",
+        values=("debug", "high", "rigorous", "safe"), multi=False)
+    variant("platform", default="nci-x86-ifort", description="Site platform",
+        values=("nci-x86-ifort", "vm-x86-gnu"), multi=False)
+
+
+    depends_on("fcm", type="build")
+    depends_on("openmpi@4.0.2:4.1.0", type=("build", "run"))
```
Comment (harshula) [inline packages/um/package.py:39]: Does openmpi need to be have a max version?

Comment (harshula) [reply packages/um/package.py:39]: When relevant use: build, link, run

File: packages/um/package.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        therefore must be statically linked.
+        """
+        env.prepend_path("LIBRARY_PATH", spec["gcom"].prefix.lib)
+
+        # Set configuration options
+        config = configparser.ConfigParser()
+        config.read(join_path(self.package_dir, "rose-app.conf"))
+        for key in config["env"]:
+            if len(key) > 0 and key[0] != '!':
+                value = config["env"][key].replace("\n=", "\n")
```
Comment (harshula) [inline packages/um/package.py:78]: Please add explanation.

File: packages/um/package.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        # Override some specific environment variables
+        env.set("optimisation_level", spec.variants["optim"].value)
+        env.set("platform_config_dir", spec.variants["platform"].value)
+        env.set("um_rev", f"vn{spec.version}")
+        components = ["casim", "jules", "shumlib", "socrates", "ukca"]
+        for comp in components:
+            key = f"{comp}_rev"
+            value = f"um{spec.version}"
+            env.set(key, value)
```
Comment (harshula) [inline packages/um/package.py:89]: Is it possible to do in one line?

File: packages/um/package.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        dep = {
+                "drhook": "drhook",
+                "eccodes": "eccodes",
+                "netcdf": "netcdf-fortran"}
+        extra_ld_flags = {
+                "drhook": "-ldrhook",
+                "eccodes": "-leccodes_f90 -leccodes",
+                "netcdf": "-lnetcdff -lnetcdf"}
```
Comment (harshula) [inline packages/um/package.py:97]: Merge into a single directionary.

File: packages/um/package.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        for libname in ["eccodes", "netcdf"]:
+            linker_args = " ".join([
+                    extra_ld_flags[libname],
+                    self._get_linker_args(spec, dep[libname])])
```
Comment (harshula) [inline packages/um/package.py:101]: Could all this be completed in `_get_linker_args()`?

File: packages/fcm/package.py (hunk 1 above)
Comment (penguian) [reply packages/fcm/package.py]: Line 28 is now removed.

File: packages/um/package.py (hunk 3 above)
Comment (penguian) [reply packages/um/package.py:92]: Added an explanation.

File: packages/um/package.py (hunk 4 above)
Comment (penguian) [reply packages/um/package.py]: Changed to one line.

File: packages/um/package.py (hunk 5 above)
Comment (penguian) [reply packages/um/package.py]: Merged into the `_lib_cfg` dictionary.

File: packages/um/package.py (hunk 6 above)
Comment (penguian) [reply packages/um/package.py]: Merged into `_get_linker_args()`

File: packages/um/package.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    def _get_linker_args(self, spec, fcm_libname):
+        """
+        The reason for the explicit -rpath is:
+        https://github.com/ACCESS-NRI/spack_packages/issues/14#issuecomment-1653651447
+        """
+        dep_name = self._lib_cfg[fcm_libname]["dep_name"]
+        ld_flags = spec[dep_name].libs.ld_flags
+        fcm_ld_flags = self._lib_cfg[fcm_libname]["fcm_ld_flags"]
+        rpaths = ["-Wl,-rpath=" + d for d in spec[dep_name].libs.directories]
+        return " ".join([ld_flags, fcm_ld_flags] + rpaths)
```
Comment (harshula) [inline packages/um/package.py:68]: Check if there's a syntax error.

Comment (penguian) [reply packages/um/package.py:68]: There is no syntax error. All check have passed: `ld_flags` and `fcm_ld_flags` are both strings, while `rpaths` is a list of strings.

**Comment by penguian** [comment]
I checked the resulting `um-atmos` and `um-recon` executables from
```
/g/data/tm70/pcl851/src/ACCESS-NRI/release/linux-rocky8-x86_64_v4/intel-2021.10.0/um-13.0-cw3uqo7aibxxsymfv7nqfiqg2bowlbgf
```
by successfully running the `u-dh821` suite.

File: packages/um/package.py [hunk 8]
```diff
@@ -128,12 +134,19 @@ def build(self, spec, prefix):
...
-            install_tree(bin_dir, prefix.bin)
+            build_name = f"build-{um_exe}"
+            build_bin_dir = join_path(
+                self._build_dir(),
+                build_name,
+                "bin")
+            install_bin_dir = join_path(
+                prefix,
+                build_name,
+                "bin")
```
Comment (harshula) [inline packages/um/package.py:150]: Can this be simplified?

File: packages/gcom/package.py [hunk 9]
```diff
@@ -9,17 +9,20 @@ class Gcom(Package):
...
     version("8.0", revision=1181)
+    version("8.1", revision=1215)
+    version("8.2", revision=1251)
+    version("8.3", revision=1288)
 
     variant("mpi", default=True, description="Build with MPI")
 
     depends_on("fcm", type="build")
-    depends_on("mpi", when="+mpi")
+    depends_on("openmpi@4.1.3:", when="+mpi", type=("build", "link", "run"))
```
Comment (harshula) [inline packages/gcom/package.py:25]: Please add a reference to the document.
Also, is it possible to do `depends_on("openmpi@4.1.3:", when="+mpi ^openmpi" ...`

**Review by harshula — APPROVED** [review]
Well done!

**Pull request merged**